pkg/cli/admin/release/extract.go

```diff
@@ -349,20 +354,42 @@ func (o *ExtractOptions) Run(ctx context.Context) error {
 		}
 	}
 
-	tarEntryCallbacks := []extract.TarEntryFunc{}
+	var versionInImageConfig string
+	opts.ImageConfigCallback = func(imageConfig *dockerv1client.DockerImageConfig) {
+		if imageConfig == nil {
+			klog.Info("Cannot retrieve the version because no image configuration is provided in the image to extract")
+			return
+		}
+		if c := imageConfig.Config; c != nil {
+			if v, ok := c.Labels["io.openshift.release"]; ok {
+				klog.V(2).Infof("Retrieved the version from image configuration in the image to extract: %s", v)
+				versionInImageConfig = v
+			} else {
+				klog.Info("Cannot retrieve the version from image configuration in the image to extract")
+			}
+		}
+	}
+
+	var manifestsCallbacks []func(string, []manifest.Manifest, io.Reader, []configv1.ClusterVersionCapability) (bool, error)
 
-	var manifestErrs []error
+	var needEnabledCapabilities bool
+	inclusionConfig := manifestInclusionConfiguration{}
 	if o.ExtractManifests {
 		expectedProviderSpecKind := credRequestCloudProviderSpecKindMapping[o.Cloud]
-
-		include := func(m *manifest.Manifest) error { return nil } // default to including everything
 		if o.Included {
 			context := "connected cluster"
-			inclusionConfig := manifestInclusionConfiguration{}
+			clientVersion, reportedVersion, err := version.ExtractVersion()
+			if err != nil {
+				return err
+			}
+			if reportedVersion == "" {
+				reportedVersion = clientVersion.String()
+			}
 			if o.InstallConfig == "" {
-				inclusionConfig, err = findClusterIncludeConfig(ctx, o.RESTConfig)
+				needEnabledCapabilities = true
+				inclusionConfig, err = findClusterIncludeConfig(ctx, o.RESTConfig, reportedVersion, versionInImageConfig)
```

Comment:
I do not understand how and why `versionInImageConfig` is used here (and on L392 too). That variable can be written by the `opts.ImageConfigCallback`, and until that callback is called, it has a string zero value: empty string. But I do not see any opportunity between here (when it is used) and L358 where that callback would be actually called, so I'd say the variable must be empty string here? Am I missing something?

Reply:
This one is in the `TarEntryCallback` (when `versionInImageConfig` is used) while `versionInImageConfig` is set up in `ImageConfigCallback` which is called earlier than `TarEntryCallback`.

We could use the existing `ImageMetadataCallback` but it is called too late.
https://github.com/openshift/oc/pull/1958#discussion_r2007764478